doc/config-generic-update-rollback/SONiC_Generic_Config_Update_and_Rollback_Design.md

```diff
@@ -0,0 +1,940 @@
+
+The other challenge to updating a switch is recoverability via rollback. Rollback needs to be with minimum-disruption e.g. if reverting ACL updates DHCP should not be affected. Currently SONiC has a couple of operations that can be candidates for rollback `config load` and `config reload`.
+
+*config reload <config_db.json>* : This command clears all the contents of the ConfigDB and loads the contents of config_db.json into the ConfigDB. After that all the Docker containers  and Linux services are restarted to establish the user specified configuration state in the config_db.json file.
+
+- Pro's
+  - Assured way of affecting a configuration state change
+- Con's
+  - Brings the links down and resets the forwarding state. This operation is disruptive in nature
+  - Time consuming as it may take 2-3 minutes for all the services to come back online. The time taken may vary based on the switch CPU power.
+- Verdict
+  - Cannot be used as a rollback mechanism
+
+*config load <config_db.json>*: This command loads the contents of config_db.json into the ConfigDB. The updates made to the ConfigDB are additive in nature and thus the new configuration state is a combination of the current running state and the partial configuration state specified by the user in the config_db.json file
+
+- Pro's
+  - Quick way to add new configuration changes
+  - It does not disrupt existing service whose configuration is not being modified. So it is non-disruptive in nature
+- Con's
+  - Can't remove existing configuration and can only be used to add/modify the existing configuration
+- Verdict
+  - Cannot be used as a rollback mechanism
+
+Since both `config load` and `config reload` are not suitable for a minimum-disruption rollback, we have to look for other approaches.
+
+In this design document, we will be exploring how to standardize the way to do partial updates, how to take checkpoints and finally how to rollback the configurations.
+
+In summary, this is the flow of an update:
+
+<img src="files/basic-target-design.png" alt="basic-target-design" width="500px"/>
+
+And the steps would be:
+```
+admin@sonic:~$ config checkpoint mycheckpoint
+admin@sonic:~$ echo "config changes to apply to ConfigDb" > some-config-changes.json-patch
+admin@sonic:~$ config apply-patch ./some-config-changes.json-patch
+admin@sonic:~$ config rollback mycheckpoint # in case of failures
+```
+
+## 1.1 Requirements
+
+### 1.1.1 Functional Requirements
+- A single, simple command to partially update SONiC configuration according to a patch of updates
+- A single, simple command to take a checkpoint of the full current SONiC config
+- A single, simple command to fully rollback current SONiC configs with to a checkpoint
+- [low-priority] A single simple command to fully replace current SONiC configs with a full config provided by an external user.
```

Comment:
> [low-priority] A single simple command to fully replace current SONiC configs with a full config provided by an external user. [](http://example.com/codeflow?start=1&length=127)

When investigating gnmi interface, we found another function requirement: gnmi SetRequest could contains multiple `delete`, `replace`, `update` operations, for example two `replace` operations, one under the path under /PORT, one under the path /PORTCHANNEL.

ref: [https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-specification.md#343-transactions](https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-specification.md#343-transactions)

The "fully replace" could be a special case for above SetRequest  requirement. As a gnmi MVP implementation, I believe supporting multiple `replace` operations will be the most high priority requirement.

I also think a CLI to replace several ConfigDB tables are handy.

Reply:
GNMI input can be converted to either:
- JsonPatch format and then we call `config apply-patch`, or
- Full target ConfigDB then we call `config replace`

We should let GCU decide the correct order of execution.